airflow/contrib/operators/gcp_function_operator.py

```diff
@@ -0,0 +1,571 @@
+
+
+def _validate_max_instances(value):
+    if int(value) <= 0:
+        raise FieldValidationException(
+            "The max instances parameter has to be greater than 0")
+
+
+LABEL_KEY_REGEXP = r'^[\w_-]{0,62}$'
+LABEL_VALUE_REGEXP = r'^[\w_-]{0,63}$'
+
+
+def _validate_labels_dict(the_dictionary):
+    # type: (dict) -> None
+    if not isinstance(the_dictionary, dict):
+        raise FieldValidationException("The field should be dictionary!")
+    for key in the_dictionary.keys():
+        if not re.match(LABEL_KEY_REGEXP, key):
+            raise FieldValidationException("The key '{}' does not match regexp '{}'".
+                                           format(key, LABEL_KEY_REGEXP))
+        value = the_dictionary[key]
+        if not re.match(LABEL_VALUE_REGEXP, value):
+            raise FieldValidationException(
+                "The value '{}' of key '{}' does not match regexp '{}'".
+                format(value, key, LABEL_VALUE_REGEXP))
+
+
+ENV_VARS_KEY_REGEXP = r'^[^\x00=]+$'
+ENV_VARS_VALUE_REGEXP = r'^[^\\x00]*$'
+
+
+def _validate_env_vars_dict(the_dictionary):
+    # type: (dict) -> None
+    if not isinstance(the_dictionary, dict):
+        raise FieldValidationException("The field should be dictionary!")
+    for key in the_dictionary.keys():
+        if key is None:
+            raise FieldValidationException("The key '{}' cannot be None".format(key))
+        if not re.match(ENV_VARS_KEY_REGEXP, key):
+            raise FieldValidationException("The key '{}' does not match regexp '{}'".
+                                           format(key, ENV_VARS_KEY_REGEXP))
+        value = the_dictionary[key]
+        if not re.match(ENV_VARS_VALUE_REGEXP, value):
+            raise FieldValidationException(
+                "The value '{}' of key '{}' does not match regexp '{}'".
+                format(value, key, ENV_VARS_VALUE_REGEXP))
+
+
+CLOUD_FUNCTION_VALIDATION = [
+    dict(name="name", regexp=r'^[^\s]+$'),
+    dict(name="description", regexp="^.+$", optional=True),
+    dict(name="entryPoint", optional=True, regexp=r'^\w+$'),
+    dict(name="runtime", regexp=r'^nodejs6$|^nodejs8$|^python37$', optional=True),
+    dict(name="timeout", regexp=r'^\d+(\.\d{1,9})?s$', optional=True),
+    dict(name="availableMemoryMb", custom_validation=_validate_available_memory_in_mb,
+         optional=True),
+    dict(name="labels", custom_validation=_validate_labels_dict, optional=True),
+    dict(name="environmentVariables", custom_validation=_validate_env_vars_dict,
+         optional=True),
+    dict(name="network",
+         regexp=r'^[^\s]+$',
+         optional=True),
+    dict(name="maxInstances", optional=True, custom_validation=_validate_max_instances),
+
+    dict(name="source_code", type="union", fields=[
+        dict(name="sourceArchiveUrl", regexp=r'^gs://[^\s]+$'),
+        dict(name="sourceRepositoryUrl", regexp=r'^gs://[^\s]+$', api='v1beta2'),
```

Comment:
Took a second look at this doc: https://cloud.google.com/functions/docs/migrating
Beta API can be changed more frequently than expected. Though unlikely, but if 'v1beta3' api also contains 'sourceRepositoryUrl' field this won't work.

Reply:
I think as it is is pretty good, I think. because what it actually checks now is that if api is v1beta3 then sourceRepositoryUrl. I changed the behaviour of union validation to simply log a warning in case none of the union fields are present in body. No exception will be raised, and validation will pass. So in case someone ads 'v2beta1' for example (I assume there will never be 'v1beta3 any more as we already have 'v1') with the `sourceRepositoryUrl` field it will continue to work, just warning will be logged. Which I think is quite acceptable. 

I will remove gs:// validation indeed. Your explanation makes perfect sense.

Reply:
Fixed in cb0bba61097c9c6a45538392015a4394f70e8905